<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2022-08-04" />

<title>Module 11: DE Visualizations and Annotation</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link rel="shortcut icon" href="favicon-16x16.png" />
<!--
Favicon from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RNA-Seq Demystified</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Sequence QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module01_Warming_Up.html">Warming up</a>
    </li>
    <li>
      <a href="Module02a_QC.html">Quality Control</a>
    </li>
    <li>
      <a href="Module02b_Cutadapt_MoreQC.html">Cutadapt/MoreQC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module03_Reference_Genomes.html">Reference Genomes</a>
    </li>
    <li>
      <a href="Module04_Alignment.html">Alignment/Quantification (STAR/RSEM)</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Quantification
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module05a_MultiQC_and_Count_Matrix.html">Count matrix / Alignment QC</a>
    </li>
    <li>
      <a href="Module05b_Additional_Details.html">Optional: Additional details</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Groups/Model
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module06_DEAnalysisSetup.html">Introduction to DESeq2</a>
    </li>
    <li>
      <a href="Module07_DESeq2Init.html">DESeq2 Initialization</a>
    </li>
    <li>
      <a href="Module08_DESeq2DE.html">Count Normalizations and DESeq2 model fitting</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Expression plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module09_SampleQCViz.html">Sample Visualizations for QC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Diff Expression
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module10_DEComparisons.html">Comparisons</a>
    </li>
    <li>
      <a href="Module11_DEVisualizations.html">DE Visiualization and Annotation</a>
    </li>
    <li>
      <a href="Module11X_BonusContent.html">Bonus content</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Module 11: DE Visualizations and
Annotation</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2022-08-04</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
<blockquote>
<h1 id="objectives" class="unlisted unnumbered">Objectives</h1>
<ul>
<li>Understand advantages of using gene ids when analyzing data.</li>
<li>Given a list of ENSEMBL gene ids, add gene symbols and Entrez
accessions.</li>
<li>Generate common visualizations for differential expression
comparisons</li>
<li>Understand choosing thresholds for calling differentially expressed
genes</li>
<li>Discuss options for functional enrichments</li>
</ul>
</blockquote>
<div id="differential-expression-workflow" class="section level1">
<h1>Differential Expression Workflow</h1>
<p>Here we will generate summary figures for our results and annotate
our DE tables.</p>
<p><img src="images/wayfinder/wayfinder-11.png"
style="width:75.0%" /></p>
<p><strong>TODO: double check references to annotated table prior to
running annotations</strong></p>
<hr />
</div>
<div id="visualizing-de-results" class="section level1">
<h1>Visualizing DE results</h1>
<p>Part of differential expression analysis is generating visualizations
to share our results. While the DESeq2 tutorial provides <a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#exploring-and-exporting-results">examples
of other visualizations</a>, a common visualization to summarize DE
comparisons are <a
href="http://resources.qiagenbioinformatics.com/manuals/clcgenomicsworkbench/752/index.php?manual=Volcano_plots_inspecting_result_statistical_analysis.html">volcano
plots</a>.</p>
<div id="generating-volcano-plots" class="section level2">
<h2>Generating Volcano plots</h2>
<p>As described by this <a
href="https://galaxyproject.github.io/training-material/topics/transcriptomics/tutorials/rna-seq-viz-with-volcanoplot/tutorial.html">Galaxy
project tutorial</a>, a volcano plot is a type of scatterplot that shows
statistical significance (adjusted p-value) versus magnitude of change
(fold change). In a volcano plot, the most upregulated genes are towards
the right, the most downregulated genes are towards the left, and the
most statistically significant genes are towards the top.</p>
<p>First, we need to set thresholds for determining significant genes. A
reasonable threshold would be a fold-change of less than -1.5 or greater
than 1.5 and an adjusted pvalue less than 0.05.</p>
<pre class="r"><code>fc = 1.5
fdr = 0.05</code></pre>
<p>Next, we’ll set up an object on which we’ll use as the basis for our
plot. Let’s coerce the <code>DataFrame</code> which was returned into a
<code>tibble</code> and use some <code>tidyverse</code> functions to do
this:</p>
<pre class="r"><code># Let&#39;s first swing the rownames out as a column and coerce to a tibble
results_plus_vs_minus = as_tibble(results_plus_vs_minus, rownames = &#39;id&#39;) %&gt;% arrange(padj)

# Now let&#39;s save a copy to add columns for our plots
df = results_plus_vs_minus

comparison = &#39;plus_vs_minus&#39;</code></pre>
<p>Once we’ve setup the annotations we need, we can proceed with
generating the volcano plot, including our thresholds.</p>
<pre class="r"><code># Initialize the plot, saving as object &#39;p&#39; and specifying the plot type as &#39;geom_point&#39;
p = ggplot(df, aes(x = log2FoldChange, y = -log10(padj))) +
    geom_point(shape = 21, colour= &#39;darkgrey&#39;, size = 1) +
    theme_bw() +
    xlab(&#39;Log2 fold-change&#39;) + ylab(&#39;-Log10 adjusted p-value&#39;)

# Add threshold lines
p = p +
    geom_vline(
        xintercept = c(0, -log2(fc), log2(fc)),
        linetype = c(1, 2, 2)) +
    geom_hline(
        yintercept = -log10(fdr),
        linetype = 2)

# Add Title that includes comparison name
p = p + ggtitle(as.character(comparison))

# Display the plot in the plots pane of RStudio
p</code></pre>
<p><img src="fig/11-VolcanoPlot-1.png" width="816" style="display: block; margin: auto;" /></p>
<details>
<summary>
<em>Click for color coded volcano plot</em>
</summary>
<p>Now we need to <a
href="https://www.statmethods.net/management/subset.html">subset our
data</a> to label the datapoints (genes) that pass our thresholds.</p>
<pre class="r"><code>df$dot = rep(3, nrow(df))
df$dot[which(df$padj &lt;= fdr &amp; df$log2FoldChange &lt; 0 &amp; abs(df$log2FoldChange) &gt;= log2(fc))] = 2
df$dot[which(df$padj &lt;= fdr &amp; df$log2FoldChange &gt; 0 &amp; abs(df$log2FoldChange) &gt;= log2(fc))] = 1
df$sig = df$dot

#take top 5 up, down, then combine, assign label
top = rbind(head(subset(df, df$dot == 1), 5),head(subset(df, df$dot == 2), 5))
top$label = top$id
df = merge(x = df, y = top[,c(&#39;id&#39;,&#39;label&#39;)], by = &quot;id&quot;, all.x = TRUE)

#count the number of significan up and down genes, assign value for legend
df$dot = factor(df$dot,levels = c(1,2,3), labels = c(paste0(&#39;Up: &#39;, sum(df$dot == 1)),paste0(&#39;Down: &#39;, sum(df$dot == 2)),&#39;NS&#39;))</code></pre>
<p>Once we’ve setup the annotations we need, we can proceed with
generating the volcano plot.</p>
<pre class="r"><code>p = ggplot(df, aes(x = log2FoldChange, y = -log10(padj))) +
    geom_point(aes(color = df$dot), size = 1) +
    theme_classic() +
    xlab(&#39;Log2 fold-change&#39;) + ylab(&#39;-Log10 adjusted p-value&#39;)
# specify colors for up-/down-/nonsignificant genes
p = p + scale_color_manual(name = &#39;&#39;, values=c(&#39;#B31B21&#39;, &#39;#1465AC&#39;, &#39;darkgray&#39;))
# add threshol lines
p = p +
    geom_vline(
        xintercept = c(0, -log2(fc), log2(fc)),
        linetype = c(1, 2, 2),
        color = c(&#39;black&#39;, &#39;black&#39;, &#39;black&#39;)) +
    geom_hline(
        yintercept = -log10(fdr),
        linetype = 2,
        color = &#39;black&#39;)
# Add Title that includes comparison name
p = p + ggtitle(as.character(comparison))

p</code></pre>
<img src="fig/11-VolcanoPlot2-1.png" width="816" style="display: block; margin: auto;" />
</details>
<p><br></p>
<p>For additional visualizations for our DE results, we included some
example code in the Bonus Content module and this <a
href="https://hbctraining.github.io/DGE_workshop/lessons/06_DGE_visualizing_results.html">HBC
tutorial</a> includes some nice examples.</p>
</div>
<div id="summarizing-our-de-results" class="section level2">
<h2>Summarizing our DE results</h2>
<p>To generate a general summary of the DE results, we can use the
<code>summary</code> function to generate a basic summary by DESeq2.</p>
<p>However, we can also use conditional statements to determine the
number of genes that pass our thresholds for each comparison, which
might be more informative.</p>
<p><em>[Question]: How would we identify the number of genes with
adjusted p-values &lt; 0.05 and a fold-change above 1.5 (or below
-1.5)?</em></p>
<pre class="r"><code>#summary(results_plus_vs_minus)
sum(results_plus_vs_minus$padj &lt; fdr &amp; abs(results_plus_vs_minus$log2FoldChange) &gt;= log2(fc), na.rm = TRUE)</code></pre>
<pre><code>[1] 56</code></pre>
<p>How do the number of DE genes compare to what we observed from the
PCA plots?</p>
<div id="choosing-thresholds" class="section level4">
<h4>Choosing thresholds</h4>
<p>Thresholding on adjusted p-values &lt; 0.05 is a standard threshold,
but depending on the research question and/or how the results will be
used, other thresholds might be reasonable.</p>
<p>There is a nice <a href="https://www.biostars.org/p/209118/">Biostar
post that discusses choosing adjusted p-value thresholds</a>, including
cases where a more relaxed threshold might be appropriate and (some
heated) discussion of the dangers of adjusting the choosen threshold
after running an analysis. Additionally, there is a <a
href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-13-S2-S11">Dalmon
et al 2012</a> paper about p-value and fold-change thresholds for
microarray data that may help provide some context.</p>
</div>
<div id="subsetting-significant-genes" class="section level3">
<h3>Subsetting significant genes</h3>
<p>It can be useful to annotate our results table with additional
information to make them easier to interpret, such as adding additional
gene information or better summarizing our DE results.</p>
<details>
<summary>
<em>Click for code to generate a table of only DE genes</em>
</summary>
<p>You may be interested in identifying only the genes that pass your
significance thresholds. A useful way to do this is to conditionally
subset your results.</p>
<p><em>Note: The tidyverse functions you learned in Software Carpentry
could also be alternatively used here.</em></p>
<pre class="r"><code>res_sig = na.omit(results_plus_vs_minus)
res_sig = res_sig[which(res_sig$padj &lt; fdr &amp; abs(res_sig$log2FoldChange) &gt;= log2(fc)), ]
head(res_sig)</code></pre>
<p>For more details about sub-setting tables in R, we recommend
reviewing the <a
href="https://datacarpentry.org/R-ecology-lesson/03-dplyr.html">Data
Carpentry manipulating and analyzing data module</a>.</p>
We can also annotate our results to include a column that identifies our
significant genes.
</details>
<p>We can also add an annotation column to indicate if a gene passes the
<code>padj</code> and <code>log2FoldChange</code> thresholds we set to
call DE genes.</p>
<pre class="r"><code># add a column and assign all genes are non-significance
results_plus_vs_minus$Call = rep(FALSE, length(results_plus_vs_minus$baseMean))

# change &#39;Call&#39; column to TRUE if meets conditions for significant differences
results_plus_vs_minus[which(!is.na(results_plus_vs_minus$padj) &amp; results_plus_vs_minus$padj &lt; 0.05 &amp; abs(results_plus_vs_minus$log2FoldChange) &gt;= log2(1.5)), ]$Call = TRUE

# reorder table to rank significant genes at the top
results_plus_vs_minus = results_plus_vs_minus[order(-results_plus_vs_minus$Call),]
head(results_plus_vs_minus)</code></pre>
</div>
</div>
<div id="adding-genome-annotations" class="section level2">
<h2>Adding genome annotations</h2>
<p>Since, gene symbols can change over time or be ambiguous we use and
recommend using the EMSEMBL reference genome and ENSEMBL ids for
alignmentsand we’ve been working with tables and data where all genes
are labeled only by their long ENSEML ID. However, this can make it
difficult to quickly look at results for genes of interest.</p>
<div id="gene-symbols-versus-gene-ids" class="section level3">
<h3>Gene symbols versus Gene IDs</h3>
<p>Luckily, Bioconductor provides many tools and resources to facilitate
access to <a
href="http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html">genomic
annotation resources</a>.</p>
<p>To start, we will first load the <a
href="https://bioconductor.org/packages/3.14/bioc/html/biomaRt.html">biomaRt
library</a> &amp; choose what reference we want to access. For a more
detailed walk through of using biomaRt, <a
href="https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2019/RNAseq/html/05_Annotation_and_Visualisation.html">this
training module</a> might be useful, including what to do when
annotations are not 1:1 mappings.</p>
<blockquote>
<h1 id="breakout-exercise-2---adding-gene-symbol-annotations"
class="unlisted unnumbered">[Breakout Exercise 2] - Adding gene symbol
annotations</h1>
<p>Navigate to <a href="">this</a> page, where we will begin our second
breakout exercise together, and then split into smaller groups to finish
it.</p>
</blockquote>
</div>
</div>
</div>
<div id="outputting-results-to-file" class="section level1">
<h1>Outputting results to file</h1>
<p>A key aspect of our analysis is preserving the relevant datasets for
both our records and for downstream applications, such as functional
enrichments.</p>
<p>First, we’ll setup a new directory for our output tables.</p>
<pre class="r"><code>dir.create(&quot;tables&quot;, showWarnings = FALSE)</code></pre>
<div id="de-results-table" class="section level2">
<h2>DE results table</h2>
<p>Next we’ll write out our DE results for the KD comparison to file,
since we added additional information to that table.</p>
<pre class="r"><code>write.csv(res_WT_anno,
          row.names = FALSE,
          na = &quot;.&quot;,
          file=&quot;tables/DEResults_ko.control_v_wt.control_annotated.csv&quot;)</code></pre>
<p>If we generated another comparision, we could repeat our annotations
or write the DE results directly to file.</p>
<pre class="r"><code>write.csv(results_plus_vs_minus,
          row.names = FALSE,
          na = &quot;.&quot;,
          file=&quot;tables/DEResults_ko.Tx_v_wt.Tx.csv&quot;)</code></pre>
</div>
<div id="count-tables" class="section level2">
<h2>Count tables</h2>
<p>The most relevant count tables are the raw, filtered count table that
we used as the input for our analysis and the rlog normalized count
table that we used for our quality control visualizations.</p>
<p>To output the raw counts, we will need to use the <code>counts</code>
function to access the count table from within its larger
<code>DESeqDataSet</code> object.</p>
<pre class="r"><code>write.csv(counts(dds, normalized = FALSE), file=&quot;tables/DESeq2_raw_counts.csv&quot;)</code></pre>
<p>Then we’ll output the rlog count table, using the <code>assay</code>
function to access the normalized count table from within its larger
<code>DESeqDataSet</code> object.</p>
<pre class="r"><code>write.csv(assay(rld), file=&quot;tables/DESeq2_rlogNormalized_counts.csv&quot;)</code></pre>
</div>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>In this section, we:</p>
<ul>
<li>Generated a volcano plot for our differential expression
results</li>
<li>Summarized our differential expression results</li>
<li>Discussed choosing thresholds</li>
<li>Annotated our tables of results to map gene IDs to gene symbols</li>
<li>Saved our results to file</li>
</ul>
</div>
<div id="key-takeaways" class="section level1">
<h1>Key takeaways</h1>
<p>Overall, we’ve run through most of the building blocks needed to run
a differential expression analysis and hopefully built up a better
understanding of how differential expression comparisons work,
particularly how experimental design can impact our results.</p>
<p>What to consider moving forward:</p>
<ul>
<li>How can I control for technical variation in my experimental
design?</li>
<li>How much variation is expected with a treatment group?</li>
<li>What is my RNA quality, and how can that be optimized?</li>
<li>Are there quality concerns for my sequencing data?</li>
<li>What comparisons are relevant to my biological question?</li>
<li>Are there covariates that should be considered?</li>
<li>What will a differential expression analysis tell me?</li>
</ul>
<p><strong>Let’s pause here for general questions</strong></p>
<hr />
</div>
<div id="what-can-we-do-with-our-de-results" class="section level1">
<h1>What can we do with our DE results?</h1>
<p>Now that we have our DE results, have we address the biological
question relevant to the authors of the <a
href="http://www.ncbi.nlm.nih.gov/pubmed/25464849">original paper</a>?
On the one hand, yes - we now have two tables of genes that are impacted
by change in Mov10 expression. But with two lists of genes alone, it can
be difficult to find patterns or understand broader biological
impacts.</p>
<p>What if we wanted to find out what genes were signifant in
<em>both</em> comparisons? The <code>intersect</code> function, such as
implimented as <a
href="https://www.datasciencemadesimple.com/intersect-function-r-using-dplyr-intersection-data-frames/">part
of the dplyr package</a> would be useful to identify shared significant
genes. A <a
href="https://www.r-graph-gallery.com/14-venn-diagramm.html">venn
diagram</a> could be a way to visualize these overlaps.</p>
<p>A way to determine possible <a
href="https://www.ebi.ac.uk/training-beta/online/courses/functional-genomics-ii-common-technologies-and-data-analysis-methods/biological-interpretation-of-gene-expression-data-2/">broader
biological interpretations</a> from the observed DE results, is
functional enrichments. There are many options, such as some included in
this <a
href="https://www.researchgate.net/post/How_can_I_analyze_a_set_of_DEGs_differentially_expressed_genes_to_obtain_information_from_them">discussion
thread</a>. Other common functional enrichments approaches are gene set
enrichment analysis, aka <a
href="http://software.broadinstitute.org/gsea/index.jsp">GSEA</a>,
Database for Annotation, Visualization and Integrated Discovery, aka <a
href="https://david.ncifcrf.gov/">DAVID</a>, <a
href="https://digitalinsights.qiagen.com/">Ingenity</a>, and [iPathway
Guide]</p>
<p>The University of Michigan has license and support for additional
tools, such as Cytoscape, so we recommend reaching out to staff with <a
href="https://www.lib.umich.edu/locations-and-hours/taubman-health-sciences-library/research-and-clinical-support">Taubman
Library</a> to learn more about resources that might be application
toyour research.</p>
<hr />
</div>
<div id="sources" class="section level1">
<h1>Sources</h1>
<ul>
<li>HBC DGE training module, part 1: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html</a></li>
<li>HBC DGE training module, part 2: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html</a></li>
<li>DESeq2 vignette: <a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis"
class="uri">http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis</a></li>
<li>Bioconductor Genomic Annotation resources: <a
href="http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html"
class="uri">http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html</a></li>
<li>BioMart vignette: <a
href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html"
class="uri">https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html</a></li>
</ul>
</div>
<div id="additional-resources" class="section level1">
<h1>Additional Resources</h1>
<ul>
<li>MIDAS Reproduciblity Hub: <a
href="https://midas.umich.edu/reproducibility-overview/"
class="uri">https://midas.umich.edu/reproducibility-overview/</a></li>
<li>ARC resources: <a href="https://arc-ts.umich.edu/"
class="uri">https://arc-ts.umich.edu/</a></li>
<li>Gene Set Enrichment Resources from Bioconductor: <a
href="https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/06_Gene_set_testing.nb.html"
class="uri">https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/06_Gene_set_testing.nb.html</a></li>
<li>Using HTSeq data with DESeq2: <a
href="https://angus.readthedocs.io/en/2019/diff-ex-and-viz.html"
class="uri">https://angus.readthedocs.io/en/2019/diff-ex-and-viz.html</a></li>
<li>Detailed RNA-seq analysis paper: <a
href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6096346/"
class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6096346/</a></li>
<li>Overview of RNA-seq analysis considerations: <a
href="https://academic-oup-com.proxy.lib.umich.edu/bfg/article/14/2/130/257370"
class="uri">https://academic-oup-com.proxy.lib.umich.edu/bfg/article/14/2/130/257370</a></li>
<li>Alternative overview of DESeq2, including visualizations and
functional enrichments: <a
href="http://dputhier.github.io/jgb71e-polytech-bioinfo-app/practical/rna-seq_R/rnaseq_diff_Snf2.html"
class="uri">http://dputhier.github.io/jgb71e-polytech-bioinfo-app/practical/rna-seq_R/rnaseq_diff_Snf2.html</a></li>
</ul>
<hr />
</div>
<div id="session-info" class="section level1">
<h1>Session Info</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.2.1 (2022-06-23)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Monterey 12.5

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] biomaRt_2.52.0              data.table_1.14.2           RColorBrewer_1.1-3          pheatmap_1.0.12            
 [5] ggrepel_0.9.1               dplyr_1.0.9                 tidyr_1.2.0                 ggplot2_3.3.6              
 [9] DESeq2_1.36.0               SummarizedExperiment_1.26.1 Biobase_2.56.0              MatrixGenerics_1.8.1       
[13] matrixStats_0.62.0          GenomicRanges_1.48.0        GenomeInfoDb_1.32.2         IRanges_2.30.0             
[17] S4Vectors_0.34.0            BiocGenerics_0.42.0         knitr_1.39                  rmarkdown_2.14             

loaded via a namespace (and not attached):
 [1] bitops_1.0-7           bit64_4.0.5            filelock_1.0.2         progress_1.2.2         httr_1.4.3            
 [6] tools_4.2.1            bslib_0.4.0            utf8_1.2.2             R6_2.5.1               DBI_1.1.3             
[11] colorspace_2.0-3       withr_2.5.0            prettyunits_1.1.1      tidyselect_1.1.2       curl_4.3.2            
[16] bit_4.0.4              compiler_4.2.1         cli_3.3.0              xml2_1.3.3             DelayedArray_0.22.0   
[21] labeling_0.4.2         sass_0.4.2             scales_1.2.0           genefilter_1.78.0      rappdirs_0.3.3        
[26] stringr_1.4.0          digest_0.6.29          XVector_0.36.0         pkgconfig_2.0.3        htmltools_0.5.3       
[31] dbplyr_2.2.1           highr_0.9              fastmap_1.1.0          rlang_1.0.4            rstudioapi_0.13       
[36] RSQLite_2.2.15         farver_2.1.1           jquerylib_0.1.4        generics_0.1.3         jsonlite_1.8.0        
[41] BiocParallel_1.30.3    RCurl_1.98-1.8         magrittr_2.0.3         GenomeInfoDbData_1.2.8 Matrix_1.4-1          
[46] Rcpp_1.0.9             munsell_0.5.0          fansi_1.0.3            lifecycle_1.0.1        stringi_1.7.8         
[51] yaml_2.3.5             zlibbioc_1.42.0        BiocFileCache_2.4.0    grid_4.2.1             blob_1.2.3            
[56] parallel_4.2.1         crayon_1.5.1           lattice_0.20-45        Biostrings_2.64.0      splines_4.2.1         
[61] annotate_1.74.0        hms_1.1.1              KEGGREST_1.36.3        locfit_1.5-9.6         pillar_1.8.0          
[66] geneplotter_1.74.0     codetools_0.2-18       XML_3.99-0.10          glue_1.6.2             evaluate_0.15         
[71] BiocManager_1.30.18    png_0.1-7              vctrs_0.4.1            gtable_0.3.0           purrr_0.3.4           
[76] assertthat_0.2.1       cachem_1.0.6           xfun_0.31              xtable_1.8-4           survival_3.3-1        
[81] tibble_3.1.7           AnnotationDbi_1.58.0   memoise_2.0.1          ellipsis_0.3.2        </code></pre>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
</div>

<div id="rmd-source-code">---
title: "Module 11: DE Visualizations and Annotation"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("11-")
```

> # Objectives {.unlisted .unnumbered}
> * Understand advantages of using gene ids when analyzing data.
> * Given a list of ENSEMBL gene ids, add gene symbols and Entrez accessions.
> * Generate common visualizations for differential expression comparisons
> * Understand choosing thresholds for calling differentially expressed genes
> * Discuss options for functional enrichments

```{r Modules, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(DESeq2)
library(ggplot2)
library(tidyr)
library(dplyr)
library(ggrepel)
library(RColorBrewer)
load("rdata/RunningData.RData")
```

# Differential Expression Workflow

Here we will generate summary figures for our results and annotate our DE tables.

![](./images/wayfinder/wayfinder-11.png){width=75%}

**TODO: double check references to annotated table prior to running annotations**

---

# Visualizing DE results

Part of differential expression analysis is generating visualizations to share our results. While the DESeq2 tutorial provides [examples of other visualizations](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#exploring-and-exporting-results), a common visualization to summarize DE comparisons are [volcano plots](http://resources.qiagenbioinformatics.com/manuals/clcgenomicsworkbench/752/index.php?manual=Volcano_plots_inspecting_result_statistical_analysis.html).

## Generating Volcano plots

As described by this [Galaxy project tutorial](https://galaxyproject.github.io/training-material/topics/transcriptomics/tutorials/rna-seq-viz-with-volcanoplot/tutorial.html), a volcano plot is a type of scatterplot that shows statistical significance (adjusted p-value) versus magnitude of change (fold change). In a volcano plot, the most upregulated genes are towards the right, the most downregulated genes are towards the left, and the most statistically significant genes are towards the top.

First, we need to set thresholds for determining significant genes. A reasonable threshold would be a fold-change of less than -1.5 or greater than 1.5 and an adjusted pvalue less than 0.05.

```{r PlotSetup1}
fc = 1.5
fdr = 0.05
```

Next, we'll set up an object on which we'll use as the basis for our plot. Let's coerce the `DataFrame` which was returned into a `tibble` and use some `tidyverse` functions to do this:

```{r PlotSetup2}
# Let's first swing the rownames out as a column and coerce to a tibble
results_plus_vs_minus = as_tibble(results_plus_vs_minus, rownames = 'id') %>% arrange(padj)

# Now let's save a copy to add columns for our plots
df = results_plus_vs_minus

comparison = 'plus_vs_minus'
```

Once we've setup the annotations we need, we can proceed with generating the volcano plot, including our thresholds.

```{r VolcanoPlot, message=FALSE, warning=FALSE}
# Initialize the plot, saving as object 'p' and specifying the plot type as 'geom_point'
p = ggplot(df, aes(x = log2FoldChange, y = -log10(padj))) +
    geom_point(shape = 21, colour= 'darkgrey', size = 1) +
    theme_bw() +
    xlab('Log2 fold-change') + ylab('-Log10 adjusted p-value')

# Add threshold lines
p = p +
    geom_vline(
        xintercept = c(0, -log2(fc), log2(fc)),
        linetype = c(1, 2, 2)) +
    geom_hline(
        yintercept = -log10(fdr),
        linetype = 2)

# Add Title that includes comparison name
p = p + ggtitle(as.character(comparison))

# Display the plot in the plots pane of RStudio
p
```

<details>
<summary>*Click for color coded volcano plot*</summary>
Now we need to [subset our data](https://www.statmethods.net/management/subset.html) to label the datapoints (genes) that pass our thresholds.

```{r PlotSetup2a, message=FALSE, warning=FALSE}
df$dot = rep(3, nrow(df))
df$dot[which(df$padj <= fdr & df$log2FoldChange < 0 & abs(df$log2FoldChange) >= log2(fc))] = 2
df$dot[which(df$padj <= fdr & df$log2FoldChange > 0 & abs(df$log2FoldChange) >= log2(fc))] = 1
df$sig = df$dot

#take top 5 up, down, then combine, assign label
top = rbind(head(subset(df, df$dot == 1), 5),head(subset(df, df$dot == 2), 5))
top$label = top$id
df = merge(x = df, y = top[,c('id','label')], by = "id", all.x = TRUE)

#count the number of significan up and down genes, assign value for legend
df$dot = factor(df$dot,levels = c(1,2,3), labels = c(paste0('Up: ', sum(df$dot == 1)),paste0('Down: ', sum(df$dot == 2)),'NS'))
```

Once we've setup the annotations we need, we can proceed with generating the volcano plot.

```{r VolcanoPlot2, message=FALSE, warning=FALSE}
p = ggplot(df, aes(x = log2FoldChange, y = -log10(padj))) +
    geom_point(aes(color = df$dot), size = 1) +
    theme_classic() +
    xlab('Log2 fold-change') + ylab('-Log10 adjusted p-value')
# specify colors for up-/down-/nonsignificant genes
p = p + scale_color_manual(name = '', values=c('#B31B21', '#1465AC', 'darkgray'))
# add threshol lines
p = p +
    geom_vline(
        xintercept = c(0, -log2(fc), log2(fc)),
        linetype = c(1, 2, 2),
        color = c('black', 'black', 'black')) +
    geom_hline(
        yintercept = -log10(fdr),
        linetype = 2,
        color = 'black')
# Add Title that includes comparison name
p = p + ggtitle(as.character(comparison))

p
```
</details>
<br>

For additional visualizations for our DE results, we included some example code in the Bonus Content module and this [HBC tutorial](https://hbctraining.github.io/DGE_workshop/lessons/06_DGE_visualizing_results.html) includes some nice examples.

## Summarizing our DE results

To generate a general summary of the DE results, we can use the `summary` function to generate a basic summary by DESeq2.

However, we can also use conditional statements to determine the number of genes that pass our thresholds for each comparison, which might be more informative.

*[Question]: How would we identify the number of genes with adjusted p-values < 0.05 and a fold-change above 1.5 (or below -1.5)?*

```{r StatSigGenes2}
#summary(results_plus_vs_minus)
sum(results_plus_vs_minus$padj < fdr & abs(results_plus_vs_minus$log2FoldChange) >= log2(fc), na.rm = TRUE)
```

How do the number of DE genes compare to what we observed from the PCA plots?

#### Choosing thresholds

Thresholding on adjusted p-values < 0.05 is a standard threshold, but depending on the research question and/or how the results will be used, other thresholds might be reasonable.

There is a nice [Biostar post that discusses choosing adjusted p-value thresholds](https://www.biostars.org/p/209118/), including cases where a more relaxed threshold might be appropriate and (some heated) discussion of the dangers of adjusting the choosen threshold after running an analysis. Additionally, there is a [Dalmon et al 2012](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-13-S2-S11) paper about p-value and fold-change thresholds for microarray data that may help provide some context.

### Subsetting significant genes

It can be useful to annotate our results table with additional information to make them easier to interpret, such as adding additional gene information or better summarizing our DE results.

<details>
<summary>*Click for code to generate a table of only DE genes*</summary>
You may be interested in identifying only the genes that pass your significance thresholds. A useful way to do this is to conditionally subset your results.

*Note: The tidyverse functions you learned in Software Carpentry could also be alternatively used here.*

```{r ConditionalSubset, eval=FALSE}
res_sig = na.omit(results_plus_vs_minus)
res_sig = res_sig[which(res_sig$padj < fdr & abs(res_sig$log2FoldChange) >= log2(fc)), ]
head(res_sig)
```

For more details about sub-setting tables in R, we recommend reviewing the [Data Carpentry manipulating and analyzing data module](https://datacarpentry.org/R-ecology-lesson/03-dplyr.html).

We can also annotate our results to include a column that identifies our significant genes.
</details>

We can also add an annotation column to indicate if a gene passes the `padj` and `log2FoldChange` thresholds we set to call DE genes.
```{r ConditionalAnno, eval=FALSE}
# add a column and assign all genes are non-significance
results_plus_vs_minus$Call = rep(FALSE, length(results_plus_vs_minus$baseMean))

# change 'Call' column to TRUE if meets conditions for significant differences
results_plus_vs_minus[which(!is.na(results_plus_vs_minus$padj) & results_plus_vs_minus$padj < 0.05 & abs(results_plus_vs_minus$log2FoldChange) >= log2(1.5)), ]$Call = TRUE

# reorder table to rank significant genes at the top
results_plus_vs_minus = results_plus_vs_minus[order(-results_plus_vs_minus$Call),]
head(results_plus_vs_minus)
```

## Adding genome annotations

Since, gene symbols can change over time or be ambiguous we use and recommend using the EMSEMBL reference genome and ENSEMBL ids for alignmentsand we've been working with tables and data where all genes are labeled only by their long ENSEML ID. However, this can make it difficult to quickly look at results for genes of interest.

### Gene symbols versus Gene IDs

Luckily, Bioconductor provides many tools and resources to facilitate access to [genomic annotation resources](http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html).

To start, we will first load the [biomaRt library](https://bioconductor.org/packages/3.14/bioc/html/biomaRt.html) & choose what reference we want to access. For a more detailed walk through of using biomaRt, [this training module](https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2019/RNAseq/html/05_Annotation_and_Visualisation.html) might be useful, including what to do when annotations are not 1:1 mappings.

> # [Breakout Exercise 2] - Adding gene symbol annotations {.unlisted .unnumbered}
>
> Navigate to [this]() page, where we will begin our second breakout exercise together, and then split into smaller groups to finish it.

# Outputting results to file

A key aspect of our analysis is preserving the relevant datasets for both our records and for downstream applications, such as functional enrichments.

First, we'll setup a new directory for our output tables.
```{r OutputTableDir}
dir.create("tables", showWarnings = FALSE)
```

## DE results table

Next we'll write out our DE results for the KD comparison to file, since we added additional information to that table.
```{r DEResultsOuput, eval = FALSE}
write.csv(res_WT_anno,
          row.names = FALSE,
          na = ".",
          file="tables/DEResults_ko.control_v_wt.control_annotated.csv")
```

If we generated another comparision, we could repeat our annotations or write the DE results directly to file.

```{r DEResultsOuput2, eval = FALSE}
write.csv(results_plus_vs_minus,
          row.names = FALSE,
          na = ".",
          file="tables/DEResults_ko.Tx_v_wt.Tx.csv")
```

## Count tables

The most relevant count tables are the raw, filtered count table that we used as the input for our analysis and the rlog normalized count table that we used for our quality control visualizations.

To output the raw counts, we will need to use the `counts` function to access the count table from within its larger `DESeqDataSet` object.

```{r OutputCountsRaw, eval = FALSE}
write.csv(counts(dds, normalized = FALSE), file="tables/DESeq2_raw_counts.csv")
```

Then we'll output the rlog count table, using the `assay` function to access the normalized count table from within its larger `DESeqDataSet` object.

```{r OutputCountsRlog, eval = FALSE}
write.csv(assay(rld), file="tables/DESeq2_rlogNormalized_counts.csv")
```

# Summary

In this section, we:

* Generated a volcano plot for our differential expression results
* Summarized our differential expression results
* Discussed choosing thresholds
* Annotated our tables of results to map gene IDs to gene symbols
* Saved our results to file


# Key takeaways

Overall, we've run through most of the building blocks needed to run a differential expression analysis and hopefully built up a better understanding of how differential expression comparisons work, particularly how experimental design can impact our results.

What to consider moving forward:

* How can I control for technical variation in my experimental design?
* How much variation is expected with a treatment group?
* What is my RNA quality, and how can that be optimized?
* Are there quality concerns for my sequencing data?
* What comparisons are relevant to my biological question?
* Are there covariates that should be considered?
* What will a differential expression analysis tell me?


**Let's pause here for general questions**

---

# What can we do with our DE results?

Now that we have our DE results, have we address the biological question relevant to the authors of the [original paper](http://www.ncbi.nlm.nih.gov/pubmed/25464849)? On the one hand, yes - we now have two tables of genes that are impacted by change in Mov10 expression. But with two lists of genes alone, it can be difficult to find patterns or understand broader biological impacts.

What if we wanted to find out what genes were signifant in *both* comparisons? The `intersect` function, such as implimented as [part of the dplyr package](https://www.datasciencemadesimple.com/intersect-function-r-using-dplyr-intersection-data-frames/) would be useful to identify shared significant genes. A [venn diagram](https://www.r-graph-gallery.com/14-venn-diagramm.html) could be a way to visualize these overlaps.

A way to determine possible [broader biological interpretations](https://www.ebi.ac.uk/training-beta/online/courses/functional-genomics-ii-common-technologies-and-data-analysis-methods/biological-interpretation-of-gene-expression-data-2/) from the observed DE results, is functional enrichments. There are many options, such as some included in this [discussion thread](https://www.researchgate.net/post/How_can_I_analyze_a_set_of_DEGs_differentially_expressed_genes_to_obtain_information_from_them). Other common functional enrichments approaches are gene set enrichment analysis, aka [GSEA](http://software.broadinstitute.org/gsea/index.jsp), Database for Annotation, Visualization and Integrated Discovery, aka [DAVID](https://david.ncifcrf.gov/), [Ingenity](https://digitalinsights.qiagen.com/), and [iPathway Guide]

The University of Michigan has license and support for additional tools, such as Cytoscape, so we recommend reaching out to staff with [Taubman Library](https://www.lib.umich.edu/locations-and-hours/taubman-health-sciences-library/research-and-clinical-support) to learn more about resources that might be application toyour research.

---

# Sources

* HBC DGE training module, part 1: https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html
* HBC DGE training module, part 2: https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html
* DESeq2 vignette: http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis
* Bioconductor Genomic Annotation resources: http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html
* BioMart vignette: https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html

# Additional Resources
* MIDAS Reproduciblity Hub: https://midas.umich.edu/reproducibility-overview/
* ARC resources: https://arc-ts.umich.edu/
* Gene Set Enrichment Resources from Bioconductor: https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/06_Gene_set_testing.nb.html
* Using HTSeq data with DESeq2: https://angus.readthedocs.io/en/2019/diff-ex-and-viz.html
* Detailed RNA-seq analysis paper: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6096346/
* Overview of RNA-seq analysis considerations: https://academic-oup-com.proxy.lib.umich.edu/bfg/article/14/2/130/257370
* Alternative overview of DESeq2, including visualizations and functional enrichments: http://dputhier.github.io/jgb71e-polytech-bioinfo-app/practical/rna-seq_R/rnaseq_diff_Snf2.html

---

```{r WriteOut.RData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
#Hidden code block to write out data for knitting
save.image(file = "rdata/RunningData_Full.RData")
```

# Session Info
```{r SessionInfo}
sessionInfo()
```

---

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Module11_DEVisualizations.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
