<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2022-08-04" />

<title>Module 07: DESeq2 Initialization</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link rel="shortcut icon" href="favicon-16x16.png" />
<!--
Favicon from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RNA-Seq Demystified</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Sequence QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module01_Warming_Up.html">Warming up</a>
    </li>
    <li>
      <a href="Module02a_QC.html">Quality Control</a>
    </li>
    <li>
      <a href="Module02b_Cutadapt_MoreQC.html">Cutadapt/MoreQC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module03_Reference_Genomes.html">Reference Genomes</a>
    </li>
    <li>
      <a href="Module04_Alignment.html">Alignment/Quantification (STAR/RSEM)</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Quantification
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module05a_MultiQC_and_Count_Matrix.html">Count matrix / Alignment QC</a>
    </li>
    <li>
      <a href="Module05b_Additional_Details.html">Optional: Additional details</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Groups/Model
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module06_DEAnalysisSetup.html">Introduction to DESeq2</a>
    </li>
    <li>
      <a href="Module07_DESeq2Init.html">DESeq2 Initialization</a>
    </li>
    <li>
      <a href="Module08_DESeq2DE.html">Count Normalizations and DESeq2 model fitting</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Expression plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module09_SampleQCViz.html">Sample Visualizations for QC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Diff Expression
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module10_DEComparisons.html">Comparisons</a>
    </li>
    <li>
      <a href="Module11_DEVisualizations.html">DE Visiualization and Annotation</a>
    </li>
    <li>
      <a href="Module11X_BonusContent.html">Bonus content</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Module 07: DESeq2 Initialization</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2022-08-04</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
<blockquote>
<h1 id="objectives" class="unlisted unnumbered">Objectives</h1>
<ul>
<li>‘Unblind’ our sample IDs</li>
<li>Understand possible confounding factors</li>
<li>Understand the impact of batches or additional covariates</li>
<li>Filter count table</li>
</ul>
</blockquote>
<div id="differential-expression-workflow" class="section level1">
<h1>Differential Expression Workflow</h1>
<p>Here we will proceed setting up the inputs needed to initialize
DESeq2 before testing for differential expression.</p>
<p><img src="images/wayfinder/wayfinder-07.png"
style="width:75.0%" /></p>
<hr />
</div>
<div id="deseq2-objects" class="section level1">
<h1>DESeq2 objects</h1>
<p>Bioconductor software packages often define and use custom classes
within R to store data in a way that better fits expectations around
biological data, such as illustrated below from <a
href="https://www.nature.com/articles/nmeth.3252">Huber et
al. 2015</a>.</p>
<p><img src="images/SummarizedExperiment.jpg" /></p>
<p>These custom data structures have pre-specified data slots, which
hold specific types/classes of data and therefore can be more easily
accessed by functions from the same package.</p>
<p>To create the DESeqDataSet we need two tables:</p>
<ol style="list-style-type: decimal">
<li>a count matrix (which we already loaded)</li>
<li>a table that assigns the condition labels for each sample (which we
will generate below).</li>
</ol>
</div>
<div id="sample-information" class="section level1">
<h1>Sample Information</h1>
<p>For this representative dataset, we have somewhat limited information
from <a
href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE144588">public
records</a>, but we know these samples were isolated from either
wild-type or knock-out T-cells harvested from control mice or isolated
from previously transplanted mice.</p>
<div id="generate-sample-table" class="section level2">
<h2>Generate Sample Table</h2>
<p>Our next step will be to describe the samples within our R session,
so that we make the proper comparisons with DESeq2. The first step is to
check the sample names from the count table.</p>
<pre class="r"><code>colnames(count_table)</code></pre>
<pre><code>[1] &quot;SRR7777895&quot; &quot;SRR7777896&quot; &quot;SRR7777897&quot; &quot;SRR7777898&quot; &quot;SRR7777899&quot; &quot;SRR7777900&quot;</code></pre>
<p>When we looked at our <code>count_table</code>, our samples are
blinded, e.g. the sample names don’t correspond to any of the expected
treatment groups. So we will need to specify which sample IDs connect to
which experimental conditions.</p>
<p>Since there are a large number of samples (and to increase the
reproduciblity of our code), we would generate a sample information
table in excel and exported it as a ‘.csv’ file so that it is in a
‘plain text’ format that can be easily loaded into our R session.</p>
<details>
<summary>
<em>Click for sample naming conventions</em>
</summary>
<p>A critical aspect of creating a sample sheet in excel is to avoid
using spaces or characters that have special uses in R, such as dashes
or parentheses. Simple sample group names are best. If you are unfamilar
with ‘.csv’ files or how to generate them, there are <a
href="https://www.wikihow.com/Create-a-CSV-File">tutorials</a> available
to guide you through the process.</p>
</details>
<p>We’ll load our ‘pre-made’ sample information sheet,
<code>samplesheet.csv</code>, to unblind our samples.</p>
<pre class="r"><code>samplesheet = read.table(&quot;data/samplesheet.csv&quot;,
                       sep = &quot;,&quot;,
                       header = TRUE,
                       row.names = 1)</code></pre>
<pre class="r"><code>samplesheet</code></pre>
<pre><code>           genotype condition
SRR7777895       wt      plus
SRR7777896       wt      plus
SRR7777897       wt      plus
SRR7777898       wt     minus
SRR7777899       wt     minus
SRR7777900       wt     minus</code></pre>
<p><strong>Checkpoint</strong>: <em>If you have loaded
<code>samplesheet</code>, please indicate with the green ‘check’ button.
Otherwise, please use the red ‘x’ button to have the command
repeated</em></p>
<p>How are the treatment groups encoded in the table we just loaded?</p>
<pre class="r"><code>unique(samplesheet$condition)</code></pre>
<pre><code>[1] &quot;plus&quot;  &quot;minus&quot;</code></pre>
<p>Next, we’ll format our table so that we have the appropriate data
type (an ordered <a
href="https://swcarpentry.github.io/r-novice-inflammation/12-supp-factors/">factor</a>)
for DESeq2 to recognize our treatment groups and appropriately compare
samples.</p>
<pre class="r"><code>samplesheet$condition = factor(samplesheet$condition,
                            levels = c(&#39;minus&#39;, &#39;plus&#39;))

unique(samplesheet$condition)</code></pre>
<pre><code>[1] plus  minus
Levels: minus plus</code></pre>
<p>Notice that we set the levels in a particular order. This is
important for setting the ‘Control’ group as the denominator in our
comparisons when we setup our DESeq2 model. The ‘Case’ name should be
last if there are only two groups &amp; first if there are only three
groups to generate default pair-wise comparisons.</p>
<p>Before we proceed, we need to make sure that the sample labels
(column names) in the count table match the sample information table
(row names), including the order.</p>
<p>If the sample labels don’t match, then we will see an error and need
to correct the labels prior to proceeding. Checking the sample
information table is extremely important to ensure that the correct
samples are grouped together for comparisons.</p>
<pre class="r"><code>all(colnames(count_table) == rownames(samplesheet))</code></pre>
<pre><code>[1] TRUE</code></pre>
<p>This line of code checks if both the identity and order match between
our <code>count_table</code> and our <code>samplesheet</code>. If, in
the course of using your own data, this ever returns <code>FALSE</code>,
try using the <code>match()</code> function to rearrange the columns of
<code>count_table</code> (or the rows of <code>samplesheet</code>) to
get them to match.</p>
<p><strong>Checkpoint</strong>: <em>If you your sample info check
returns <code>TRUE</code>, please indicate with the green ‘yes’ button.
Otherwise, please use the red ‘x’ button to have the command
repeated</em></p>
</div>
</div>
<div id="creating-deseq2-object" class="section level1">
<h1>Creating DESeq2 object</h1>
<p>To create the <code>DESeqDataSet</code> we will need the
<code>count_table</code> and the <code>samplesheet</code>. We will also
need a <strong>design formula</strong> to specify our model.</p>
<div id="making-model-choices" class="section level2">
<h2>Making model choices</h2>
<p>The design formula specified informs many of the DESeq2 functions how
to treat the samples in the analysis, specifically which column in the
samaple metadata table contains the experimental design.</p>
<p>In this case, we aren’t aware of any <a
href="https://methods-sagepub-com.proxy.lib.umich.edu/reference/encyc-of-research-design/n85.xml"><strong>covariates</strong></a>
that should be considered in our comparisons. However, if there are
additional attributes of the samples that may impact the DE comparisons,
like sex, date of collection, or patient of origin, these should be
added as <a href="https://support.bioconductor.org/p/75309/">additional
columns</a> in the sample information table and <a
href="https://support.bioconductor.org/p/98700/">added to a design
formula</a>.</p>
<details>
<summary>
<em>Click for Note</em>
</summary>
More complex questions, including determining if a fold-change due to
treatment is different across groups, such as patient samples,
“interaction terms” can be included in the design formula, such as
outlined in <a href="https://support.bioconductor.org/p/98628/">this
support thread</a>.
</details>
<p>The design formula specifies the column(s) in the metadata table and
how they should be used in the analysis. For our dataset we only have
one column we are interested in, that is <code>condition</code>. This
column has three factor levels, which tells DESeq2 that for each gene we
want to evaluate gene expression change with respect to these different
levels.</p>
<pre class="r"><code>## Create DESeq object, line by line
dds = DESeqDataSetFromMatrix(countData = count_table,
                              colData = samplesheet,
                              design = ~ condition)</code></pre>
<pre><code>converting counts to integer mode</code></pre>
<pre class="r"><code>dds</code></pre>
<pre><code>class: DESeqDataSet 
dim: 55492 6 
metadata(1): version
assays(1): counts
rownames(55492): ENSMUSG00000000001 ENSMUSG00000000003 ... mcherry tdtomato
rowData names(0):
colnames(6): SRR7777895 SRR7777896 ... SRR7777899 SRR7777900
colData names(2): genotype condition</code></pre>
<p>Notice that printing the <code>dds</code> object helpfully shows us
some helpful information:</p>
<ul>
<li>The dimension (number of genes by number of samples),</li>
<li>The gene identifiers,</li>
<li>The sample identifiers,</li>
<li>The additional column names giving information about the
samples</li>
</ul>
<p><strong>Checkpoint</strong>: <em>If you see <code>dds</code> in your
environment panel, please indicate with the green ‘check’ button.
Otherwise, please use use the red ‘x’ button in your zoom reaction panel
to have this step repeated. You can use the red ‘x’ to be put in a
breakout room for help</em></p>
<details>
<summary>
<em>Click for how to model batch effects with DESeq2</em>
</summary>
<p>Differences between samples can also be due to technical reasons,
such as collection on different days or different sequencing runs.
Differences between samples that are not due to biological factors as
called <strong>batch effects</strong>. We can include batch effects in
our design model in the same way as covariates, as long as the technical
groups do not overlap, or <strong>confound</strong>, the biological
treatment groups. Let’s try add some additional meta-data information
where we have counfounding batch effects and create another DESeq2
object.</p>
<pre class="r"><code>#samplesheet
samplesheet$batch = factor(c(rep(c(&quot;Day1&quot;), 3),
                       rep(c(&quot;Day2&quot;), 3),
                       rep(c(&quot;Day3&quot;), 3),
                       rep(c(&quot;Day4&quot;), 3)),
                     levels = c(&quot;Day1&quot;, &quot;Day2&quot;, &quot;Day3&quot;, &quot;Day4&quot;))

dds_batch = DESeqDataSetFromMatrix(countData = count_table,
                      colData = samplesheet,
                      design = ~ batch + condition)</code></pre>
Notice that if you run the above command, the error indicates that
variables in the design formula “are linear combinations” which means
that batch and condition are correlated and the function is unable to
fill in a required ‘slot’ in the DESeq2 object. So if batches are
<strong>not balanced</strong> by including both case and controls (like
in the patient covariate example) then we cannot control for those
technical effects through statistical modeling.
</details>
<hr />
</div>
</div>
<div id="pre-filtering" class="section level1">
<h1>Pre-filtering</h1>
<p>While not necessary, <a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pre-filtering">pre-filtering</a>
the count table helps to not only reduce the size of the DESeq2 object,
but also gives you a sense of how many genes were reasonably measured at
the sequencing depth generated for your samples.</p>
<p>Here we will filter out any genes that have less than 10 counts
across any of the samples. This is a fairly standard level of filtering,
but can filter data less/more depending on quality control metrics from
alignments and sequencing depth or total number of samples.</p>
<pre class="r"><code>keep = rowSums(counts(dds)) &gt;= 10
dds = dds[keep,]
dds</code></pre>
<pre><code>class: DESeqDataSet 
dim: 16249 6 
metadata(1): version
assays(1): counts
rownames(16249): ENSMUSG00000000001 ENSMUSG00000000028 ... ENSMUSG00000118651 ENSMUSG00000118653
rowData names(0):
colnames(6): SRR7777895 SRR7777896 ... SRR7777899 SRR7777900
colData names(2): genotype condition</code></pre>
<p>Notice the <code>dds</code> object has less elements after filtering.
There were quite a number of genes that were not measured in our
experiment.</p>
<p><strong>Checkpoint</strong>: <em>Questions?</em></p>
<div id="optional-exercise-adding-covariates-to-sample-information"
class="section level4 unlisted unnumbered">
<h4 class="unlisted unnumbered">[Optional exercise]: Adding covariates
to sample information</h4>
<p>You can test out manually adding a covariate to our
<code>samplesheet</code> data table and then create a new DESeq2
object.</p>
<details>
<summary>
<em>Click for how to add a covariate to sample information</em>
</summary>
<pre class="r"><code>head(samplesheet)
samplesheet$patient = factor(rep(c(&quot;P1&quot;, &quot;P2&quot;, &quot;P3&quot;), 4), levels = c(&quot;P1&quot;, &quot;P2&quot;, &quot;P3&quot;))
head(samplesheet)</code></pre>
<p>Notice how we avoid starting with a number our patient covariate
labels since R doesn’t like that.</p>
<pre class="r"><code>dds_patient = DESeqDataSetFromMatrix(countData = count_table,
                          colData = samplesheet,
                          design = ~ patient + condition)</code></pre>
Now we have specified for DESeq2 that we want to test for the effect of
the condition (the last factor) while controlling for the effect of the
patient origin (the first factor).
</details>
</div>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>In this section, we:</p>
<ul>
<li>Loaded the necessary input files into our R session</li>
<li>Discussed model design for DESeq2</li>
<li>Initialized a DESeq2 data set</li>
<li>Filtered our count data</li>
</ul>
<p>Now that we’ve created our DESeq2 objects, including specifying what
model is appropriate for our data, and filtered our data, we can proceed
with assessing the impact of the experimental conditions on gene
expression across our samples.</p>
<hr />
</div>
<div id="sources" class="section level1">
<h1>Sources</h1>
<p>Training resources used to develop materials:</p>
<ul>
<li>HBC DGE setup: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html</a></li>
<li>HBC Count Normalization: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html</a></li>
<li>DESeq2 standard vignette: <a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html"
class="uri">http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html</a></li>
<li>DESeq2 beginners vignette: <a
href="https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf"
class="uri">https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf</a></li>
<li>Bioconductor RNA-seq Workflows: <a
href="https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1_RNASeq.html"
class="uri">https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1_RNASeq.html</a></li>
<li>CCDL Gastric cancer training materials: <a
href="https://alexslemonade.github.io/training-modules/RNA-seq/03-gastric_cancer_exploratory.nb.html"
class="uri">https://alexslemonade.github.io/training-modules/RNA-seq/03-gastric_cancer_exploratory.nb.html</a></li>
<li>CCDL Neuroblastoma training materials: <a
href="https://alexslemonade.github.io/training-modules/RNA-seq/05-nb_cell_line_DESeq2.nb.html"
class="uri">https://alexslemonade.github.io/training-modules/RNA-seq/05-nb_cell_line_DESeq2.nb.html</a></li>
</ul>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
</div>

<div id="rmd-source-code">---
title: "Module 07: DESeq2 Initialization"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("07-")
```

> # Objectives {.unlisted .unnumbered}
> * 'Unblind' our sample IDs
> * Understand possible confounding factors
> * Understand the impact of batches or additional covariates
> * Filter count table

```{r LoadRunningData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(DESeq2)
load("rdata/RunningData.RData")
```


# Differential Expression Workflow

Here we will proceed setting up the inputs needed to initialize DESeq2 before testing for differential expression.

![](./images/wayfinder/wayfinder-07.png){width=75%}

---

# DESeq2 objects

Bioconductor software packages often define and use custom classes within R to store data in a way that better fits expectations around biological data, such as illustrated below from [Huber et al. 2015](https://www.nature.com/articles/nmeth.3252).

![](./images/SummarizedExperiment.jpg)

These custom data structures have pre-specified data slots, which hold specific types/classes of data and therefore can be more easily accessed by functions from the same package.

To create the DESeqDataSet we need two tables:

1. a count matrix (which we already loaded)
2. a table that assigns the condition labels for each sample (which we will generate below).


# Sample Information

For this representative dataset, we have somewhat limited information from [public records](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE144588), but we know these samples were isolated from either wild-type or knock-out T-cells harvested from control mice or isolated from previously transplanted mice.

## Generate Sample Table

Our next step will be to describe the samples within our R session, so that we make the proper comparisons with DESeq2. The first step is to check the sample names from the count table.
```{r ColumnNames}
colnames(count_table)
```

When we looked at our `count_table`, our samples are blinded, e.g. the sample names don't correspond to any of the expected treatment groups. So we will need to specify which sample IDs connect to which experimental conditions.

Since there are a large number of samples (and to increase the reproduciblity of our code), we would generate a sample information table in excel and exported it as a '.csv' file so that it is in a 'plain text' format that can be easily loaded into our R session.


<details>
    <summary>*Click for sample naming conventions*</summary>
    A critical aspect of creating a sample sheet in excel is to avoid using spaces or characters that have special uses in R, such as dashes or parentheses. Simple sample group names are best.
    If you are unfamilar with '.csv' files or how to generate them, there are [tutorials](https://www.wikihow.com/Create-a-CSV-File) available to guide you through the process.

</details>

We'll load our 'pre-made' sample information sheet, `samplesheet.csv`, to unblind our samples.

```{r Samplesheet, echo = FALSE, eval = TRUE}
samplesheet = read.table("../data/samplesheet.csv",
                       sep = ",",
                       header = TRUE,
                       row.names = 1)
```

```{r Samplesheet2, eval = FALSE}
samplesheet = read.table("data/samplesheet.csv",
                       sep = ",",
                       header = TRUE,
                       row.names = 1)
```

```{r ShowSamplesheet}
samplesheet
```

**Checkpoint**: *If you have loaded `samplesheet`, please indicate with the green 'check' button. Otherwise, please use the red 'x' button to have the command repeated*

How are the treatment groups encoded in the table we just loaded?

```{r TreatmentGroupTable}
unique(samplesheet$condition)
```

Next, we'll format our table so that we have the appropriate data type (an ordered [factor](https://swcarpentry.github.io/r-novice-inflammation/12-supp-factors/)) for DESeq2 to recognize our treatment groups and appropriately compare samples.

```{r SamplesheetFactor}

samplesheet$condition = factor(samplesheet$condition,
                            levels = c('minus', 'plus'))

unique(samplesheet$condition)
```

Notice that we set the levels in a particular order. This is important for setting the 'Control' group as the denominator in our comparisons when we setup our DESeq2 model. The 'Case' name should be last if there are only two groups & first if there are only three groups to generate default pair-wise comparisons.

Before we proceed, we need to make sure that the sample labels (column names) in the count table match the sample information table (row names), including the order.

If the sample labels don't match, then we will see an error and need to correct the labels prior to proceeding. Checking the sample information table is extremely important to ensure that the correct samples are grouped together for comparisons.

```{r CheckSamplesheet}
all(colnames(count_table) == rownames(samplesheet))
```
This line of code checks if both the identity and order match between our `count_table` and our `samplesheet`. If, in the course of using your own data, this ever returns `FALSE`, try using the `match()` function to rearrange the columns of `count_table` (or the rows of `samplesheet`) to get them to match.

**Checkpoint**: *If you your sample info check returns `TRUE`, please indicate with the green 'yes' button. Otherwise, please use the red 'x' button to have the command repeated*

# Creating DESeq2 object

To create the `DESeqDataSet` we will need the `count_table` and the `samplesheet`. We will also need a **design formula** to specify our model.

## Making model choices

The design formula specified informs many of the DESeq2 functions how to treat the samples in the analysis, specifically which column in the samaple metadata table contains the experimental design.

In this case, we aren't aware of any [**covariates**](https://methods-sagepub-com.proxy.lib.umich.edu/reference/encyc-of-research-design/n85.xml) that should be considered in our comparisons. However, if there are additional attributes of the samples that may impact the DE comparisons, like sex, date of collection, or patient of origin, these should be added as [additional columns](https://support.bioconductor.org/p/75309/) in the sample information table and [added to a design formula](https://support.bioconductor.org/p/98700/).


<details>
    <summary>*Click for Note*</summary>
    More complex questions, including determining if a fold-change due to treatment is different across groups, such as patient samples, "interaction terms" can be included in the design formula, such as outlined in [this support thread](https://support.bioconductor.org/p/98628/).
</details>

The design formula specifies the column(s) in the metadata table and how they should be used in the analysis. For our dataset we only have one column we are interested in, that is `condition`. This column has three factor levels, which tells DESeq2 that for each gene we want to evaluate gene expression change with respect to these different levels.

```{r DESeq2Object}
## Create DESeq object, line by line
dds = DESeqDataSetFromMatrix(countData = count_table,
                              colData = samplesheet,
                              design = ~ condition)
dds
```

Notice that printing the `dds` object helpfully shows us some helpful information:

* The dimension (number of genes by number of samples),
* The gene identifiers,
* The sample identifiers,
* The additional column names giving information about the samples

**Checkpoint**: *If you see `dds` in your environment panel, please indicate with the green 'check' button. Otherwise, please use  use the red 'x' button in your zoom reaction panel to have this step repeated. You can use the red 'x' to be put in a breakout room for help*

<details>
    <summary>*Click for how to model batch effects with DESeq2*</summary>
    Differences between samples can also be due to technical reasons, such as collection on different days or different sequencing runs. Differences between samples that are not due to biological factors as called **batch effects**. We can include batch effects in our design model in the same way as covariates, as long as the technical groups do not overlap, or **confound**, the biological treatment groups.
    Let's try add some additional meta-data information where we have counfounding batch effects and create another DESeq2 object.

```{r Confounders, eval=FALSE}
#samplesheet
samplesheet$batch = factor(c(rep(c("Day1"), 3),
                       rep(c("Day2"), 3),
                       rep(c("Day3"), 3),
                       rep(c("Day4"), 3)),
                     levels = c("Day1", "Day2", "Day3", "Day4"))

dds_batch = DESeqDataSetFromMatrix(countData = count_table,
                      colData = samplesheet,
                      design = ~ batch + condition)
```

    Notice that if you run the above command, the error indicates that variables in the design formula "are linear combinations" which means that batch and condition are correlated and the function is unable to fill in a required 'slot' in the DESeq2 object.
    So if batches are **not balanced** by including both case and controls (like in the patient covariate example) then we cannot control for those technical effects through statistical modeling.
</details>

---


# Pre-filtering

While not necessary, [pre-filtering](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pre-filtering) the count table helps to not only reduce the size of the DESeq2 object, but also gives you a sense of how many genes were reasonably measured at the sequencing depth generated for your samples.

Here we will filter out any genes that have less than 10 counts across any of the samples. This is a fairly standard level of filtering, but can filter data less/more depending on quality control metrics from alignments and sequencing depth or total number of samples.

```{r PreFilter}
keep = rowSums(counts(dds)) >= 10
dds = dds[keep,]
dds
```

Notice the `dds` object has less elements after filtering. There were quite a number of genes that were not measured in our experiment.

**Checkpoint**: *Questions?*

#### [Optional exercise]: Adding covariates to sample information {.unlisted .unnumbered}

You can test out manually adding a covariate to our `samplesheet` data table and then create a new DESeq2 object.

<details>
    <summary>*Click for how to add a covariate to sample information*</summary>

```{r Covariates, eval=FALSE}
head(samplesheet)
samplesheet$patient = factor(rep(c("P1", "P2", "P3"), 4), levels = c("P1", "P2", "P3"))
head(samplesheet)
```

    Notice how we avoid starting with a number our patient covariate labels since R doesn't like that.

```{r CovariateDESeq2Object, eval=FALSE}
dds_patient = DESeqDataSetFromMatrix(countData = count_table,
                          colData = samplesheet,
                          design = ~ patient + condition)
```

    Now we have specified for DESeq2 that we want to test for the effect of the condition (the last factor) while controlling for the effect of the patient origin (the first factor).
</details>

# Summary

In this section, we:

* Loaded the necessary input files into our R session
* Discussed model design for DESeq2
* Initialized a DESeq2 data set
* Filtered our count data

Now that we've created our DESeq2 objects, including specifying what model is appropriate for our data, and filtered our data, we can proceed with assessing the impact of the experimental conditions on gene expression across our samples.

---

# Sources

Training resources used to develop materials:

* HBC DGE setup: https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html
* HBC Count Normalization: https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html
* DESeq2 standard vignette: http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
* DESeq2 beginners vignette: https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf
* Bioconductor RNA-seq Workflows: https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1_RNASeq.html
* CCDL Gastric cancer training materials: https://alexslemonade.github.io/training-modules/RNA-seq/03-gastric_cancer_exploratory.nb.html
* CCDL Neuroblastoma training materials: https://alexslemonade.github.io/training-modules/RNA-seq/05-nb_cell_line_DESeq2.nb.html



```{r WriteOut.RData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
#Hidden code block to write out data for knitting

save.image(file = "rdata/RunningData.RData")

```

---

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Module07_DESeq2Init.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
