<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2022-08-04" />

<title>Breakout Exercise 1 - Expression Heatmap</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link rel="shortcut icon" href="favicon-16x16.png" />
<!--
Favicon from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RNA-Seq Demystified</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Sequence QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module01_Warming_Up.html">Warming up</a>
    </li>
    <li>
      <a href="Module02a_QC.html">Quality Control</a>
    </li>
    <li>
      <a href="Module02b_Cutadapt_MoreQC.html">Cutadapt/MoreQC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module03_Reference_Genomes.html">Reference Genomes</a>
    </li>
    <li>
      <a href="Module04_Alignment.html">Alignment/Quantification (STAR/RSEM)</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Quantification
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module05a_MultiQC_and_Count_Matrix.html">Count matrix / Alignment QC</a>
    </li>
    <li>
      <a href="Module05b_Additional_Details.html">Optional: Additional details</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Groups/Model
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module06_DEAnalysisSetup.html">Introduction to DESeq2</a>
    </li>
    <li>
      <a href="Module07_DESeq2Init.html">DESeq2 Initialization</a>
    </li>
    <li>
      <a href="Module08_DESeq2DE.html">Count Normalizations and DESeq2 model fitting</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Expression plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module09_SampleQCViz.html">Sample Visualizations for QC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Diff Expression
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module10_DEComparisons.html">Comparisons</a>
    </li>
    <li>
      <a href="Module11_DEVisualizations.html">DE Visiualization and Annotation</a>
    </li>
    <li>
      <a href="Module11X_BonusContent.html">Bonus content</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Breakout Exercise 1 - Expression
Heatmap</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2022-08-04</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
<p>Estimated time: <strong>15 Minutes</strong></p>
<div id="motivation" class="section level1">
<h1>Motivation</h1>
<p>Plotting the expression values across all samples for the top
variable genes in an experiment can help to visualize how samples
cluster together by their expression profiles. When combined with
phenotypic data, it can help show how samples with different treatments
behave relative to one another.</p>
</div>
<div id="exercise" class="section level1">
<h1>Exercise</h1>
<p>Create a heatmap of the top 50 most variable genes using the
<code>pheatmap()</code> function.</p>
</div>
<div id="instructions" class="section level1">
<h1>Instructions</h1>
<ul>
<li>One group member should share their screen in the breakout room. If
nobody volunteers, a helper may randomly select someone.</li>
<li>The group members should discuss the exercise and work together to
find a solution.</li>
<li>If there is time after a solution is found, allow all members to
complete the exercise.</li>
</ul>
</div>
<div id="example" class="section level1">
<h1>Example</h1>
<p>To get an idea of what we expect to see at the end, let’s look at a
toy example from the <code>pheatmap()</code> help examples. There’s no
need to run this code, we just want to illustrate the form of the
result.</p>
<pre class="r"><code># Copied from the pheatmap documentation

# Create matrix with random normally distributed values
test = matrix(rnorm(200), 20, 10)

# Impose some structure so the heatmap appears interesting
test[1:10, seq(1, 10, 2)] = test[1:10, seq(1, 10, 2)] + 3
test[11:20, seq(2, 10, 2)] = test[11:20, seq(2, 10, 2)] + 2
test[15:20, seq(2, 10, 2)] = test[15:20, seq(2, 10, 2)] + 4

# Name the rows and columns
colnames(test) = paste(&quot;Test&quot;, 1:10, sep = &quot;&quot;)
rownames(test) = paste(&quot;Gene&quot;, 1:20, sep = &quot;&quot;)

# Draw the heatmap
pheatmap(test)</code></pre>
<p><img src="fig/09a-test_heatmap-1.png" width="816" style="display: block; margin: auto;" /></p>
</div>
<div id="steps" class="section level1">
<h1>Steps</h1>
<ol style="list-style-type: decimal">
<li>Look at the documentation for the <code>pheatmap()</code> function
and determine what the most important parameter is. Hint: It’s usually
the first or first few parameters. This is telling us what we’ll have to
get for the next step.</li>
</ol>
<details>
<summary>
Answer
</summary>
<pre class="r"><code>?pheatmap</code></pre>
For this exercise, we’ll only need to use the <code>mat</code>
parameter, giving the numeric matrix to be plotted.
</details>
<p><br></p>
<ol start="2" style="list-style-type: decimal">
<li>Extract the rlog normalized expression values for the experiment.
Hint: We created an <code>rld</code> object earlier. The
<code>assay()</code> function pulls out the values.</li>
</ol>
<details>
<summary>
Answer
</summary>
<pre class="r"><code>exp_mat = assay(rld)
head(exp_mat)</code></pre>
<pre><code>                   SRR7777895 SRR7777896 SRR7777897 SRR7777898 SRR7777899 SRR7777900
ENSMUSG00000000001  10.514813  10.366709  10.419463  10.840373  10.410449  10.578771
ENSMUSG00000000028  10.604461  10.734506  10.735026  10.682714  10.820938  10.990999
ENSMUSG00000000031  11.160276  10.498747  10.742763  11.861617  10.578156  10.298022
ENSMUSG00000000037   4.642433   4.555500   4.578934   4.656950   4.719012   4.620204
ENSMUSG00000000049   3.017478   2.820198   2.843894   2.936455   2.814833   2.811619
ENSMUSG00000000056  14.321672  14.284652  14.337197  13.904238  14.393912  14.235971</code></pre>
</details>
<p><br></p>
<ol start="3" style="list-style-type: decimal">
<li>Calculate the variance for each gene in the expression matrix we
just extracted. Hint: Look at the help for
<code>matrixStats::rowVars()</code> and decide if that’s a reasonable
function to use.</li>
</ol>
<details>
<summary>
Answer
</summary>
<pre class="r"><code>gene_vars = rowVars(exp_mat)
head(gene_vars)</code></pre>
<pre><code>[1] 0.030341701 0.017700141 0.326567308 0.003409940 0.007135949 0.030863320</code></pre>
</details>
<p><br></p>
<ol start="4" style="list-style-type: decimal">
<li>Get the numerical indices for the top 50 most variable genes. Hint:
Run the <code>order()</code> function on a toy example, like
<code>order(c(-1.25, 1.3, 5.6, 2.1))</code>, and think about what is
being returned. Note, it’s not the values in the original vector. Then
look at the help for <code>order()</code> and figure out how how to
reverse what is returned.</li>
</ol>
<details>
<summary>
Answer
</summary>
<pre class="r"><code>order(c(-1.25, 1.3, 5.6, 2.1))</code></pre>
<pre><code>[1] 1 2 4 3</code></pre>
<pre class="r"><code>order(c(-1.25, 1.3, 5.6, 2.1), decreasing = TRUE)</code></pre>
<pre><code>[1] 3 4 2 1</code></pre>
<pre class="r"><code>ordered_idx = order(gene_vars, decreasing = TRUE)
top_50_idx = ordered_idx[1:50]</code></pre>
</details>
<p><br></p>
<ol start="5" style="list-style-type: decimal">
<li>Subset the expression matrix from step 2 using this index vector.
Hint: Remember square-bracket notation, and that we want to subset the
rows, while returning all the columns. Make sure the result has the
number of rows you expect, that is, 50.</li>
</ol>
<details>
<summary>
Answer
</summary>
<pre class="r"><code>top_var_exp_mat = exp_mat[top_50_idx, ]
dim(top_var_exp_mat)</code></pre>
<pre><code>[1] 50  6</code></pre>
</details>
<p><br></p>
<ol start="6" style="list-style-type: decimal">
<li>Create a heatmap using this subsetted expression matrix using the
<code>pheatmap()</code> function.</li>
</ol>
<details>
<summary>
Answer
</summary>
<pre class="r"><code>pheatmap(top_var_exp_mat)</code></pre>
<img src="fig/09a-create_heatmap-1.png" width="816" style="display: block; margin: auto;" />
</details>
<p><br></p>
</div>
<div id="saving-the-result" class="section level1">
<h1>Saving the result</h1>
<p>If time permits, discuss with your group how you might save this
heatmap. Hint: Look at the parameters for the function in
<code>?pheatmap</code>. Alternatively, consider how we saved the PCA in
the previous module.</p>
</div>

<div id="rmd-source-code">LS0tCnRpdGxlOiAiQnJlYWtvdXQgRXhlcmNpc2UgMSAtIEV4cHJlc3Npb24gSGVhdG1hcCIKYXV0aG9yOiAiVU0gQmlvaW5mb3JtYXRpY3MgQ29yZSIKZGF0ZTogImByIFN5cy5EYXRlKClgIgpvdXRwdXQ6CiAgICAgICAgaHRtbF9kb2N1bWVudDoKICAgICAgICAgICAgaW5jbHVkZXM6CiAgICAgICAgICAgICAgICBpbl9oZWFkZXI6IGhlYWRlci5odG1sCiAgICAgICAgICAgIHRoZW1lOiBwYXBlcgogICAgICAgICAgICB0b2M6IHRydWUKICAgICAgICAgICAgdG9jX2RlcHRoOiA0CiAgICAgICAgICAgIHRvY19mbG9hdDogdHJ1ZQogICAgICAgICAgICBudW1iZXJfc2VjdGlvbnM6IGZhbHNlCiAgICAgICAgICAgIGZpZ19jYXB0aW9uOiB0cnVlCiAgICAgICAgICAgIG1hcmtkb3duOiBHRk0KICAgICAgICAgICAgY29kZV9kb3dubG9hZDogdHJ1ZQotLS0KCjxzdHlsZSB0eXBlPSJ0ZXh0L2NzcyI+CmJvZHksIHRkIHsKICAgZm9udC1zaXplOiAxOHB4Owp9CmNvZGUucnsKICBmb250LXNpemU6IDEycHg7Cn0KcHJlIHsKICBmb250LXNpemU6IDEycHgKfQo8L3N0eWxlPgoKYGBge3IsIGluY2x1ZGUgPSBGQUxTRX0Kc291cmNlKCIuLi9iaW4vY2h1bmstb3B0aW9ucy5SIikKa25pdHJfZmlnX3BhdGgoIjA5YS0iKQpgYGAKCmBgYHtyIE1vZHVsZXMsIGV2YWw9VFJVRSwgZWNobz1GQUxTRSwgbWVzc2FnZT1GQUxTRSwgd2FybmluZz1GQUxTRX0KbGlicmFyeShERVNlcTIpCmxpYnJhcnkoZ2dwbG90MikKbGlicmFyeSh0aWR5cikKbGlicmFyeShkcGx5cikKbGlicmFyeShtYXRyaXhTdGF0cykKbGlicmFyeShnZ3JlcGVsKQpsaWJyYXJ5KHBoZWF0bWFwKQpsaWJyYXJ5KFJDb2xvckJyZXdlcikKbG9hZCgicmRhdGEvUnVubmluZ0RhdGEuUkRhdGEiKQpgYGAKCkVzdGltYXRlZCB0aW1lOiAqKjE1IE1pbnV0ZXMqKgoKIyBNb3RpdmF0aW9uCgpQbG90dGluZyB0aGUgZXhwcmVzc2lvbiB2YWx1ZXMgYWNyb3NzIGFsbCBzYW1wbGVzIGZvciB0aGUgdG9wIHZhcmlhYmxlIGdlbmVzIGluIGFuIGV4cGVyaW1lbnQgY2FuIGhlbHAgdG8gdmlzdWFsaXplIGhvdyBzYW1wbGVzIGNsdXN0ZXIgdG9nZXRoZXIgYnkgdGhlaXIgZXhwcmVzc2lvbiBwcm9maWxlcy4gV2hlbiBjb21iaW5lZCB3aXRoIHBoZW5vdHlwaWMgZGF0YSwgaXQgY2FuIGhlbHAgc2hvdyBob3cgc2FtcGxlcyB3aXRoIGRpZmZlcmVudCB0cmVhdG1lbnRzIGJlaGF2ZSByZWxhdGl2ZSB0byBvbmUgYW5vdGhlci4KCiMgRXhlcmNpc2UKCkNyZWF0ZSBhIGhlYXRtYXAgb2YgdGhlIHRvcCA1MCBtb3N0IHZhcmlhYmxlIGdlbmVzIHVzaW5nIHRoZSBgcGhlYXRtYXAoKWAgZnVuY3Rpb24uCgojIEluc3RydWN0aW9ucwoKLSBPbmUgZ3JvdXAgbWVtYmVyIHNob3VsZCBzaGFyZSB0aGVpciBzY3JlZW4gaW4gdGhlIGJyZWFrb3V0IHJvb20uIElmIG5vYm9keSB2b2x1bnRlZXJzLCBhIGhlbHBlciBtYXkgcmFuZG9tbHkgc2VsZWN0IHNvbWVvbmUuCi0gVGhlIGdyb3VwIG1lbWJlcnMgc2hvdWxkIGRpc2N1c3MgdGhlIGV4ZXJjaXNlIGFuZCB3b3JrIHRvZ2V0aGVyIHRvIGZpbmQgYSBzb2x1dGlvbi4KLSBJZiB0aGVyZSBpcyB0aW1lIGFmdGVyIGEgc29sdXRpb24gaXMgZm91bmQsIGFsbG93IGFsbCBtZW1iZXJzIHRvIGNvbXBsZXRlIHRoZSBleGVyY2lzZS4KCiMgRXhhbXBsZQoKVG8gZ2V0IGFuIGlkZWEgb2Ygd2hhdCB3ZSBleHBlY3QgdG8gc2VlIGF0IHRoZSBlbmQsIGxldCdzIGxvb2sgYXQgYSB0b3kgZXhhbXBsZSBmcm9tIHRoZSBgcGhlYXRtYXAoKWAgaGVscCBleGFtcGxlcy4gVGhlcmUncyBubyBuZWVkIHRvIHJ1biB0aGlzIGNvZGUsIHdlIGp1c3Qgd2FudCB0byBpbGx1c3RyYXRlIHRoZSBmb3JtIG9mIHRoZSByZXN1bHQuCgpgYGB7ciB0ZXN0X2hlYXRtYXB9CiMgQ29waWVkIGZyb20gdGhlIHBoZWF0bWFwIGRvY3VtZW50YXRpb24KCiMgQ3JlYXRlIG1hdHJpeCB3aXRoIHJhbmRvbSBub3JtYWxseSBkaXN0cmlidXRlZCB2YWx1ZXMKdGVzdCA9IG1hdHJpeChybm9ybSgyMDApLCAyMCwgMTApCgojIEltcG9zZSBzb21lIHN0cnVjdHVyZSBzbyB0aGUgaGVhdG1hcCBhcHBlYXJzIGludGVyZXN0aW5nCnRlc3RbMToxMCwgc2VxKDEsIDEwLCAyKV0gPSB0ZXN0WzE6MTAsIHNlcSgxLCAxMCwgMildICsgMwp0ZXN0WzExOjIwLCBzZXEoMiwgMTAsIDIpXSA9IHRlc3RbMTE6MjAsIHNlcSgyLCAxMCwgMildICsgMgp0ZXN0WzE1OjIwLCBzZXEoMiwgMTAsIDIpXSA9IHRlc3RbMTU6MjAsIHNlcSgyLCAxMCwgMildICsgNAoKIyBOYW1lIHRoZSByb3dzIGFuZCBjb2x1bW5zCmNvbG5hbWVzKHRlc3QpID0gcGFzdGUoIlRlc3QiLCAxOjEwLCBzZXAgPSAiIikKcm93bmFtZXModGVzdCkgPSBwYXN0ZSgiR2VuZSIsIDE6MjAsIHNlcCA9ICIiKQoKIyBEcmF3IHRoZSBoZWF0bWFwCnBoZWF0bWFwKHRlc3QpCmBgYAoKIyBTdGVwcwoKMS4gTG9vayBhdCB0aGUgZG9jdW1lbnRhdGlvbiBmb3IgdGhlIGBwaGVhdG1hcCgpYCBmdW5jdGlvbiBhbmQgZGV0ZXJtaW5lIHdoYXQgdGhlIG1vc3QgaW1wb3J0YW50IHBhcmFtZXRlciBpcy4gSGludDogSXQncyB1c3VhbGx5IHRoZSBmaXJzdCBvciBmaXJzdCBmZXcgcGFyYW1ldGVycy4gVGhpcyBpcyB0ZWxsaW5nIHVzIHdoYXQgd2UnbGwgaGF2ZSB0byBnZXQgZm9yIHRoZSBuZXh0IHN0ZXAuCgo8ZGV0YWlscz4KPHN1bW1hcnk+QW5zd2VyPC9zdW1tYXJ5PgpgYGB7ciBwaGVhdG1hcF9oZWxwLCBldmFsID0gRkFMU0V9Cj9waGVhdG1hcApgYGAKCkZvciB0aGlzIGV4ZXJjaXNlLCB3ZSdsbCBvbmx5IG5lZWQgdG8gdXNlIHRoZSBgbWF0YCBwYXJhbWV0ZXIsIGdpdmluZyB0aGUgbnVtZXJpYyBtYXRyaXggdG8gYmUgcGxvdHRlZC4KPC9kZXRhaWxzPgo8YnI+CgoyLiBFeHRyYWN0IHRoZSBybG9nIG5vcm1hbGl6ZWQgZXhwcmVzc2lvbiB2YWx1ZXMgZm9yIHRoZSBleHBlcmltZW50LiBIaW50OiBXZSBjcmVhdGVkIGFuIGBybGRgIG9iamVjdCBlYXJsaWVyLiBUaGUgYGFzc2F5KClgIGZ1bmN0aW9uIHB1bGxzIG91dCB0aGUgdmFsdWVzLgoKPGRldGFpbHM+CjxzdW1tYXJ5PkFuc3dlcjwvc3VtbWFyeT4KYGBge3IgZXh0cmFjdF9ybGRfZXhwcmVzc2lvbn0KZXhwX21hdCA9IGFzc2F5KHJsZCkKaGVhZChleHBfbWF0KQpgYGAKPC9kZXRhaWxzPgo8YnI+CgozLiBDYWxjdWxhdGUgdGhlIHZhcmlhbmNlIGZvciBlYWNoIGdlbmUgaW4gdGhlIGV4cHJlc3Npb24gbWF0cml4IHdlIGp1c3QgZXh0cmFjdGVkLiBIaW50OiBMb29rIGF0IHRoZSBoZWxwIGZvciBgbWF0cml4U3RhdHM6OnJvd1ZhcnMoKWAgYW5kIGRlY2lkZSBpZiB0aGF0J3MgYSByZWFzb25hYmxlIGZ1bmN0aW9uIHRvIHVzZS4KCjxkZXRhaWxzPgo8c3VtbWFyeT5BbnN3ZXI8L3N1bW1hcnk+CmBgYHtyIGNhbGNfcm93X3ZhcnN9CmdlbmVfdmFycyA9IHJvd1ZhcnMoZXhwX21hdCkKaGVhZChnZW5lX3ZhcnMpCmBgYAo8L2RldGFpbHM+Cjxicj4KCjQuIEdldCB0aGUgbnVtZXJpY2FsIGluZGljZXMgZm9yIHRoZSB0b3AgNTAgbW9zdCB2YXJpYWJsZSBnZW5lcy4gSGludDogUnVuIHRoZSBgb3JkZXIoKWAgZnVuY3Rpb24gb24gYSB0b3kgZXhhbXBsZSwgbGlrZSBgb3JkZXIoYygtMS4yNSwgMS4zLCA1LjYsIDIuMSkpYCwgYW5kIHRoaW5rIGFib3V0IHdoYXQgaXMgYmVpbmcgcmV0dXJuZWQuIE5vdGUsIGl0J3Mgbm90IHRoZSB2YWx1ZXMgaW4gdGhlIG9yaWdpbmFsIHZlY3Rvci4gVGhlbiBsb29rIGF0IHRoZSBoZWxwIGZvciBgb3JkZXIoKWAgYW5kIGZpZ3VyZSBvdXQgaG93IGhvdyB0byByZXZlcnNlIHdoYXQgaXMgcmV0dXJuZWQuCgo8ZGV0YWlscz4KPHN1bW1hcnk+QW5zd2VyPC9zdW1tYXJ5PgpgYGB7ciBvcmRlcl92YXJzfQpvcmRlcihjKC0xLjI1LCAxLjMsIDUuNiwgMi4xKSkKb3JkZXIoYygtMS4yNSwgMS4zLCA1LjYsIDIuMSksIGRlY3JlYXNpbmcgPSBUUlVFKQoKb3JkZXJlZF9pZHggPSBvcmRlcihnZW5lX3ZhcnMsIGRlY3JlYXNpbmcgPSBUUlVFKQp0b3BfNTBfaWR4ID0gb3JkZXJlZF9pZHhbMTo1MF0KYGBgCjwvZGV0YWlscz4KPGJyPgoKNS4gU3Vic2V0IHRoZSBleHByZXNzaW9uIG1hdHJpeCBmcm9tIHN0ZXAgMiB1c2luZyB0aGlzIGluZGV4IHZlY3Rvci4gSGludDogUmVtZW1iZXIgc3F1YXJlLWJyYWNrZXQgbm90YXRpb24sIGFuZCB0aGF0IHdlIHdhbnQgdG8gc3Vic2V0IHRoZSByb3dzLCB3aGlsZSByZXR1cm5pbmcgYWxsIHRoZSBjb2x1bW5zLiBNYWtlIHN1cmUgdGhlIHJlc3VsdCBoYXMgdGhlIG51bWJlciBvZiByb3dzIHlvdSBleHBlY3QsIHRoYXQgaXMsIDUwLgoKPGRldGFpbHM+CjxzdW1tYXJ5PkFuc3dlcjwvc3VtbWFyeT4KYGBge3IgZXh0cmFjdF90b3BfZXhwfQp0b3BfdmFyX2V4cF9tYXQgPSBleHBfbWF0W3RvcF81MF9pZHgsIF0KZGltKHRvcF92YXJfZXhwX21hdCkKYGBgCjwvZGV0YWlscz4KPGJyPgoKNi4gQ3JlYXRlIGEgaGVhdG1hcCB1c2luZyB0aGlzIHN1YnNldHRlZCBleHByZXNzaW9uIG1hdHJpeCB1c2luZyB0aGUgYHBoZWF0bWFwKClgIGZ1bmN0aW9uLgoKPGRldGFpbHM+CjxzdW1tYXJ5PkFuc3dlcjwvc3VtbWFyeT4KYGBge3IgY3JlYXRlX2hlYXRtYXB9CnBoZWF0bWFwKHRvcF92YXJfZXhwX21hdCkKYGBgCjwvZGV0YWlscz4KPGJyPgoKIyBTYXZpbmcgdGhlIHJlc3VsdAoKSWYgdGltZSBwZXJtaXRzLCBkaXNjdXNzIHdpdGggeW91ciBncm91cCBob3cgeW91IG1pZ2h0IHNhdmUgdGhpcyBoZWF0bWFwLiBIaW50OiBMb29rIGF0IHRoZSBwYXJhbWV0ZXJzIGZvciB0aGUgZnVuY3Rpb24gaW4gYD9waGVhdG1hcGAuIEFsdGVybmF0aXZlbHksIGNvbnNpZGVyIGhvdyB3ZSBzYXZlZCB0aGUgUENBIGluIHRoZSBwcmV2aW91cyBtb2R1bGUuCgpgYGB7ciBXcml0ZU91dC5SRGF0YSwgZXZhbD1GQUxTRSwgZWNobz1GQUxTRSwgbWVzc2FnZT1GQUxTRSwgd2FybmluZz1GQUxTRX0KI0hpZGRlbiBjb2RlIGJsb2NrIHRvIHdyaXRlIG91dCBkYXRhIGZvciBrbml0dGluZwpzYXZlLmltYWdlKGZpbGUgPSAicmRhdGEvUnVubmluZ0RhdGEuUkRhdGEiKQpgYGAK</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Module09a_breakout.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
